File: getting_started/61.旋转链表.py
#
# @lc app=leetcode.cn id=61 lang=python3
#
# [61] 旋转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k == 0:
            return head
        if head is None:
            return head
        n = 0
        pointer = head
        while pointer.next is not None:
            n += 1
            pointer = pointer.next
        if n == 0:
            return head
        pointer.next = head
        k = k % (n+1)
        print(k)
        new_pointer = head
        for i in range(n-k):
            new_pointer = new_pointer.next
               
        result = new_pointer.next
        new_pointer.next = None
        return result
        

        
# @lc code=end

File: getting_started/61.旋转链表.cpp
/*
 * @lc app=leetcode.cn id=61 lang=cpp
 *
 * [61] 旋转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {

        if(k == 0)
        {
            return head;
        }
        if(head == nullptr)
        {
            return head;
        }
        
        ListNode *temp_node = head;
        ListNode *pre;
        int n = 0;
        while (temp_node != nullptr)
        {
            pre = temp_node;
            temp_node = temp_node->next;
            n ++;
        }
        
        k = k % n;
        if(k == 0)
        {
            return head;
        }

        pre->next = head;
        int i = 0;
        temp_node = head;

        while(i + k <n)
        {
            i ++;
            pre = temp_node;
            temp_node = temp_node->next;
        }
        pre->next = nullptr;
        return temp_node;
        
    }
};
// @lc code=end

